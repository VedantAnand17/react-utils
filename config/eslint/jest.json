{
  "env": {
    "browser": true,
    "jest": true
  },
  "plugins": [
    "jest"
  ],
  "rules": {
    "@typescript-eslint/no-implied-eval": 0,
    "global-require": 0,
    "import/no-dynamic-require": 0,
    "import/no-extraneous-dependencies": 0,
    "no-console": 0,
    "no-new-func": 0,
    "react/require-default-props": 0
  }
}