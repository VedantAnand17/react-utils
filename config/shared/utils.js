/**
 * Misc utilities reusable in different JS-format config files.
 */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const path = require('path');
const cssesc = require('cssesc');

const { interpolateName } = require('loader-utils');

/**
 * Normalizes file path to OS-independent format (adopted from css-loader).
 * @ignore
 * @param {string} file
 * @return {string}
 */
function normalizePath(file) {
  return path.sep === '\\' ? file.replace(/\\/g, '/') : file;
}

// eslint-disable-next-line no-control-regex
const filenameReservedRegex = /[<>:"/\\|?*]/g;
// eslint-disable-next-line no-control-regex
const reControlChars = /[\u0000-\u001f\u0080-\u009f]/g;

function escapeLocalident(localident) {
  return cssesc(
    localident
      // For `[hash]` placeholder
      .replace(/^((-?[0-9])|--)/, '_$1')
      .replace(filenameReservedRegex, '-')
      .replace(reControlChars, '-')
      .replace(/\./g, '-'),
    { isIdentifier: true },
  );
}

/**
 * Returns the name of package containing the folder; i.e. it recursively looks
 * up from the folder for the closest package.json file, and returns the name in
 * that file. It also caches the results from previously fisited folders.
 * @ignore
 * @param {string} folder
 * @return {string}
 */
function getPackageInfo(folder) {
  let res = getPackageInfo.cache[folder];
  if (!res) {
    const p = path.resolve(folder, 'package.json');
    /* eslint-disable global-require, import/no-dynamic-require */
    res = fs.existsSync(p) ? {
      name: require(p).name,
      root: folder,
    } : getPackageInfo(path.resolve(folder, '..'));
    /* eslint-enable global-require, import/no-dynamic-require */
    getPackageInfo.cache[folder] = res;
  }
  return res;
}
getPackageInfo.cache = {};

function getLocalIdent(
  { resourcePath },
  localIdentName,
  localName,
  options = {},
) {
  const packageInfo = getPackageInfo(path.dirname(resourcePath));
  const request = normalizePath(path.relative(packageInfo.root, resourcePath));
  return interpolateName({
    resourcePath,
  }, localIdentName, {
    ...options,
    content: `${packageInfo.name + request}\x00${localName}`,
    context: packageInfo.root,
  }).replace(/\[package\]/gi, packageInfo.name)
    .replace(/\[local\]/gi, localName)
    .replace(/@/g, '-');
}

function generateScopedNameDev(localName, assetPath) {
  // Mind: "css-loader" additionally escapes identifiers returned by
  // getLocalIdent(); the babel plugins - don't, thus we need to escape
  // ourselves.
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[package]___[path][name]___[local]___[hash:base64:6]',
      localName,
      {},
    ),
  );
}

function generateScopedNameProd(localName, assetPath) {
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[hash:base64:6]',
      localName,
      {},
    ),
  );
}

module.exports = {
  generateScopedNameDev,
  generateScopedNameProd,
  getLocalIdent,
};
