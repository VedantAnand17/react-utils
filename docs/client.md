# Client

**TODO: To be revised! Does not seem to match the actual implementation!**

Client-side intialization code.

- **Why?** &mdash; For the same reasons we have the standard
  [server setup](./server.md).

This module exports a single **`Launch(options)`** function that starts the app
at the client side. It renders the app, wraps it with
[`react-router`](https://www.npmjs.com/package/react-router)'s `<BrowserRouter>`,
and Redux store `<Provider>`, if necessary; then injects it into the page HTML
template generated by the server. In the case of use with development server,
it also takes care about HMR (Hot Module Reloading) setup at the clients-side.

**`options`** should be a JS object, with the following fields:
- **`Application`** &mdash; *Function* &mdash; The root ReactJS component of
  the app to be rendered.

  **Arguments**
  - **`initialState`** &mdash; *Object* &mdash; Initial global state, injected by
    the server (will be `undefined`, if it is not generated by the current
    server setup).

  **Returns**
  - *Promise* that resolves to the created Redux store.
