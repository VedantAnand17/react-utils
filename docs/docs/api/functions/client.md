# client()
```jsx
import { client } from '@dr.pogodin/react-utils';

client(Application, options);
```
Initializes and launches ReactJS `Application` at the client-side.

Namely, it wraps the app with [BrowserRouter] and [GlobalStateProvider],
then hydrates it into DOM element with **react-view** ID, assuming
[the standard SSR setup] was used to generate the initial HTML markup of the page.

The `dontHydrate` option allows to opt for a pure client-side rendering, instead
of the hydration.

## Arguments
- `Application` - **React.Component** - The root applciation component.
- `options` - **object** - Optional. Additional settings:
  - `dontHydrate` - **boolean** - By default, the app is hydrated into DOM
    element found by **react-view** ID, using React's [hydrateRoot()] functionality,
    and  assuming [the standard SSR setup] was used to generate the initial HTML markup of the page.

    With `dontHydrate` flag set **true** React's [createRoot()] function is used
    instead to perform a clean client-side rendering into the DOM element with
    **react-view** ID.

<!-- links -->
[BrowserRouter]: https://reactrouter.com/docs/en/v6/api#browserrouter
[GlobalStateProvider]: https://dr.pogodin.studio/docs/react-global-state/docs/api/components/globalstateprovider
[the standard SSR setup]: /docs/api/functions/server
[createRoot()]: https://reactjs.org/docs/react-dom-client.html#createroot
[hydrateRoot()]: https://reactjs.org/docs/react-dom-client.html#hydrateroot