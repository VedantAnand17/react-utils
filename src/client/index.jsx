// Initialization of client-side code.
/* global document, window */

import { GlobalStateProvider } from '@dr.pogodin/react-global-state';

import ReactDom from 'react-dom';
import { BrowserRouter } from 'react-router-dom';

/**
 * @category Utilities
 * @func client
 * @desc
 * ```js
 * import { client } from '@dr.pogodin/react-utils';
 * ```
 * Prepares and launches the app at client side.
 * @param {object} Application Root application component
 */
export default function Launch(Application) {
  ReactDom.hydrate((
    <GlobalStateProvider initialState={window.ISTATE}>
      <BrowserRouter>
        <Application />
      </BrowserRouter>
    </GlobalStateProvider>
  ), document.getElementById('react-view'));
}
