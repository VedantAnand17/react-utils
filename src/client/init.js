/**
 * Initialization of client-side environment.
 */

/* global BUILD_INFO, document, window */

// Note: this way, only required part of "node-forge": AES, and some utils,
// is bundled into client-side code.
import forge from 'node-forge/lib/forge';
import 'node-forge/lib/aes';

/* `BUILD_INFO` is always injected by Webpack build, but this check is needed
 * to adopt the code for usage in tests. */
if (typeof BUILD_INFO !== 'undefined') {
  window.TRU_BUILD_INFO = BUILD_INFO;
}

/* Removes data injection script out of the document.
 * The if block is here for test purposes. */
if (!window.TRU_KEEP_INJ_SCRIPT) {
  const block = document.querySelector('script[id="inj"]');
  document.getElementsByTagName('body')[0].removeChild(block);
}

/* TODO: A proper logger should be moved to `@dr.pogodin/react-utils`. */
/* eslint-disable no-console */
const { useServiceWorker } = window.TRU_BUILD_INFO;
if (useServiceWorker) {
  const { navigator } = window;
  if ('serviceWorker' in navigator) {
    window.addEventListener('load', async () => {
      try {
        const reg = await navigator
          .serviceWorker.register('/__service-worker.js');
        console.log('SW registered:', reg);
      } catch (err) {
        console.log('SW registration failed:', err);
      }
    });
  }
}
/* eslint-enable no-console */

/* Decodes data injected at the server side. */
const { key } = window.TRU_BUILD_INFO;
let data = forge.util.decode64(window.INJ);
const decipher = forge.cipher.createDecipher('AES-CBC', key);
decipher.start({ iv: data.slice(0, 32) });
decipher.update(forge.util.createBuffer(data.slice(32)));
decipher.finish();
data = JSON.parse(forge.util.decodeUtf8(decipher.output.data));

window.CHUNK_GROUPS = data.CHUNK_GROUPS;
window.CONFIG = data.CONFIG;
window.ISTATE = data.ISTATE;
