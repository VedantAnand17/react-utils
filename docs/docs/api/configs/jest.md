# Jest
Default configuration for [Jest] testing framework.

:::caution Beware
It does strong assumptions about the host project's code structure: the host
[Jest] configuration in the `/config/jest` folder of host code; the source code
in `/src` folder; build outputs in `/build` folder; _etc._

For a project with different structure it might be easier to compose
the appropriate Jest config from scratch.
:::

To use this default configuration create two files in the `/config/jest` folder
of host project:

```js title="/config/jest/config.js"
const config = require('@dr.pogodin/react-utils/config/jest/default');

// You can tune config as necessary here.

module.exports = config;
```

```js title="/config/jest/setup.js"
import '@dr.pogodin/react-utils/config/jest/setup';
```

After that point to that host config file from the Jest execution command,
_e.g._:
```bash
jest --config config/jest/config.js
```

[Jest]: https://jestjs.io
