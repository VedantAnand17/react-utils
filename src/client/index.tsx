// Initialization of client-side code.
/* global document */

import { GlobalStateProvider } from '@dr.pogodin/react-global-state';

import { createRoot, hydrateRoot } from 'react-dom/client';
import { BrowserRouter } from 'react-router-dom';

import getInj from './getInj';

/**
 * Prepares and launches the app at client side.
 * @param {object} Application Root application component
 * @param {object} [options={}] Optional. Additional settings.
 */
export default function Launch(Application, options = {}) {
  const container = document.getElementById('react-view');
  const scene = (
    <GlobalStateProvider initialState={getInj().ISTATE}>
      <BrowserRouter>
        <Application />
      </BrowserRouter>
    </GlobalStateProvider>
  );

  if (options.dontHydrate) {
    const root = createRoot(container);
    root.render(scene);
  } else hydrateRoot(container, scene);
}
