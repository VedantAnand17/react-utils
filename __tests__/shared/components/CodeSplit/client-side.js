/** @jest-environment jsdom */

import { Helmet } from 'react-helmet';
import mockdate from 'mockdate';
import pretty from 'pretty';

import { act, mockClientSide } from 'utils/jest';

import { renderServerSide } from './__assets__/shared';

jest.mock('node-forge');
jest.mock('uuid');

jest.useFakeTimers();
mockdate.set('2019-11-29Z');

test('Client-side rendering', async () => {
  // Emulates server-sider render to figure out the markup that should be
  // matched at the client-side.

  const { IS_SERVER_SIDE } = require('utils/isomorphy');
  expect(IS_SERVER_SIDE).toBe(true);

  // Forces Helmet to believe it runs server-side.
  Helmet.canUseDOM = false;

  let SampleCodeSplit = require('./__assets__/SampleCodeSplit').default;
  let serverMarkup = renderServerSide(SampleCodeSplit, 3);
  await jest.runAllTimers();
  serverMarkup = await serverMarkup;
  document.write(serverMarkup);
  document.close();
  const ssrHead = pretty(document.head.innerHTML);
  const ssrBody = pretty(document.body.innerHTML);
  expect(ssrHead).toMatchSnapshot();
  expect(ssrBody).toMatchSnapshot();

  // Forces Helmet to believe it runs client-side.
  Helmet.canUseDOM = true;

  // Environment reset to client-side emulation, and the actual test.

  jest.resetModules();
  mockClientSide();
  window.TRU_KEEP_INJ_SCRIPT = true;
  SampleCodeSplit = require('./__assets__/SampleCodeSplit').default;
  require('client/init');

  const { IS_CLIENT_SIDE } = require('utils/isomorphy');
  expect(IS_CLIENT_SIDE).toBe(true);

  let Launch = require('client').default;
  act(() => Launch({ getApplication: () => SampleCodeSplit }));
  let head = pretty(document.head.innerHTML);
  let body = pretty(document.body.innerHTML);
  expect(head).toEqual(ssrHead);
  expect(body).toEqual(ssrBody);

  /* This tests that in the real render the data injection script
   * is auto-removed from the document during the injection. */
  jest.resetModules();
  SampleCodeSplit = require('./__assets__/SampleCodeSplit').default;
  document.write(serverMarkup);
  document.close();
  delete window.TRU_KEEP_INJ_SCRIPT;
  require('client/init');
  Launch = require('client').default;
  act(() => Launch({ getApplication: () => SampleCodeSplit }));
  head = pretty(document.head.innerHTML);
  body = pretty(document.body.innerHTML);
  expect(head).toMatchSnapshot();
  expect(body).toMatchSnapshot();
});
