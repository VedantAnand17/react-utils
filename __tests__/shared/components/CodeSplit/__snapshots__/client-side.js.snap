// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client-side rendering 1`] = `
"<title data-react-helmet=\\"true\\"></title>

<meta name=\\"theme-color\\" content=\\"#FFFFFF\\">

<link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\">
<link href=\\"/test/public/path/sample-component-1-1511941200000.css\\" rel=\\"stylesheet\\">
<link href=\\"/test/public/path/sample-component-2-1511941200000.css\\" rel=\\"stylesheet\\">
<link href=\\"/test/public/path/sample-component-3-1511941200000.css\\" rel=\\"stylesheet\\">

<meta charset=\\"utf-8\\">
<meta content=\\"width=device-width,initial-scale=1.0\\" name=\\"viewport\\">"
`;

exports[`Client-side rendering 2`] = `
"<div id=\\"react-view\\">
  <h1>testValue</h1>
  <div data-chunk-name=\\"sample-component-1\\">
    <div>
      <h1>Sample Component</h1><pre>{
  \\"test\\": {
    \\"key\\": \\"testValue\\"
  },
  \\"dr_pogodin_react_utils___split_components\\": {
    \\"sample-component-1\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    },
    \\"sample-component-2\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    }
  }
}</pre>
    </div>
  </div>
  <div data-chunk-name=\\"sample-component-2\\">
    <div>
      <h1>Sample Component</h1><pre>{
  \\"test\\": {
    \\"key\\": \\"testValue\\"
  },
  \\"dr_pogodin_react_utils___split_components\\": {
    \\"sample-component-1\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    },
    \\"sample-component-2\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    }
  }
}</pre>
    </div>
  </div>
  <div data-chunk-name=\\"sample-component-3\\">
    <div>
      <h1>Sample Component</h1><pre>{
  \\"test\\": {
    \\"key\\": \\"testValue\\"
  },
  \\"dr_pogodin_react_utils___split_components\\": {
    \\"sample-component-1\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    },
    \\"sample-component-2\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    }
  }
}</pre>
    </div>
  </div>
</div>
<script id=\\"inj\\" type=\\"application/javascript\\" nonce=\\"abcdef-dummy-nonce\\">
  window.INJ = \\"MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghsfwaFoaus6qLoke9MlgqF39wUyiJOuxKmqjECLiFel3AfVxBlTkA2sj6FNYLq9mBb1VDFWdXsViFyeX13tuynQ88EETx6RuQyms6aHDjsZzm54u6pdtE5bElWjYQQYYKWQ6Q7AG6jOJ0m4jyHJEL2hbRtKnBKIGH+RZ8NtSPyZUdQIHOOyAmEjNWwTcoflKwqX1yZQFqY61t71j0sQev32jJQxdFCAfnIPEBbQyoMOthnONT2MDoAhYBo7mW5WvIIwKZOdbtEV00hm8UfXRtHfiDZlOuY93yXGAlvaXY18BqM3DPSi818K7Z1H4SWC1VYTGrNkXoMDIE6ybStZpZjR4spwNSQv32yICqmIFSe5L6VP07D+zuaRT8l8Kisyjw5b2Xh3ZZWsXzZjicu32lyebon9uckaiJy3Jj8fVO83oGnoVoG23K8ksWQe5vTPxUvDNJMayc7CzFtcUZFtdicQT0RPhKYtcvngtzc19MY6rKHS5iMFX5Ms6j+LMWfYk3Np31TfQ3q5XhqvVHZPciRN1TTzEdcfcEJ+d8oQ+6GKM=\\"
</script>
<script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
<script src=\\"/test/public/path/sample-component-2-1511941200000.js\\" type=\\"application/javascript\\"></script>
<script src=\\"/test/public/path/sample-component-3-1511941200000.js\\" type=\\"application/javascript\\"></script>"
`;

exports[`Client-side rendering 3`] = `
"<title data-react-helmet=\\"true\\"></title>

<meta name=\\"theme-color\\" content=\\"#FFFFFF\\">

<link href=\\"/test/public/path/main-1511941200000.css\\" rel=\\"stylesheet\\">
<link href=\\"/test/public/path/sample-component-1-1511941200000.css\\" rel=\\"stylesheet\\">
<link href=\\"/test/public/path/sample-component-2-1511941200000.css\\" rel=\\"stylesheet\\">
<link href=\\"/test/public/path/sample-component-3-1511941200000.css\\" rel=\\"stylesheet\\">

<meta charset=\\"utf-8\\">
<meta content=\\"width=device-width,initial-scale=1.0\\" name=\\"viewport\\">"
`;

exports[`Client-side rendering 4`] = `
"<div id=\\"react-view\\">
  <h1>testValue</h1>
  <div data-chunk-name=\\"sample-component-1\\">
    <div>
      <h1>Sample Component</h1><pre>{
  \\"test\\": {
    \\"key\\": \\"testValue\\"
  },
  \\"dr_pogodin_react_utils___split_components\\": {
    \\"sample-component-1\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    },
    \\"sample-component-2\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    }
  }
}</pre>
    </div>
  </div>
  <div data-chunk-name=\\"sample-component-2\\">
    <div>
      <h1>Sample Component</h1><pre>{
  \\"test\\": {
    \\"key\\": \\"testValue\\"
  },
  \\"dr_pogodin_react_utils___split_components\\": {
    \\"sample-component-1\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    },
    \\"sample-component-2\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    }
  }
}</pre>
    </div>
  </div>
  <div data-chunk-name=\\"sample-component-3\\">
    <div>
      <h1>Sample Component</h1><pre>{
  \\"test\\": {
    \\"key\\": \\"testValue\\"
  },
  \\"dr_pogodin_react_utils___split_components\\": {
    \\"sample-component-1\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    },
    \\"sample-component-2\\": {
      \\"data\\": {},
      \\"numRefs\\": 0,
      \\"operationId\\": \\"\\",
      \\"timestamp\\": 1574985600000
    }
  }
}</pre>
    </div>
  </div>
</div>

<script src=\\"/test/public/path/main-1511941200000.js\\" type=\\"application/javascript\\"></script>
<script src=\\"/test/public/path/sample-component-2-1511941200000.js\\" type=\\"application/javascript\\"></script>
<script src=\\"/test/public/path/sample-component-3-1511941200000.js\\" type=\\"application/javascript\\"></script>"
`;
